# Part 1
### Use the Notebook to build the code to scrape the following Wikipedia page, https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M.

In [1]:
import pandas as pd

In [2]:
url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
data = pd.read_html(url, header=0)
type(data)

list

In [3]:
df1 = data[0]
df1.head()

Postal Code           Borough               Neighborhood
0         M1A      Not assigned                        NaN
1         M2A      Not assigned                        NaN
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

In [4]:
df1.shape

(180, 3)

In [5]:
df1 = df1[df1.Borough != 'Not assigned']

In [6]:
df1.head()

Postal Code           Borough                                 Neighborhood
2         M3A        North York                                    Parkwoods
3         M4A        North York                             Victoria Village
4         M5A  Downtown Toronto                    Regent Park, Harbourfront
5         M6A        North York             Lawrence Manor, Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [7]:
df1.shape

(103, 3)

# Part 2
### Use the Geocoder package or the csv file to create the following dataframe:

In [8]:
df2 = pd.read_csv('Geospatial_Coordinates.csv')
df2.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [9]:
df2.shape

(103, 3)

In [10]:
neighborhoods = pd.merge(df1, df2, how='left', on='Postal Code')
neighborhoods.head()

Postal Code           Borough                                 Neighborhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

In [11]:
neighborhoods.shape

(103, 5)

# Part 3

## STEP 1. Explore Dataset

First of all, let's make sure that the dataset has all 10 boroughs and 103 neighborhoods.

In [12]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(neighborhoods['Borough'].unique()),
        neighborhoods.shape[0]
    )
)

The dataframe has 10 boroughs and 103 neighborhoods.


In [13]:
print(neighborhoods['Borough'].unique())

['North York' 'Downtown Toronto' 'Etobicoke' 'Scarborough' 'East York'
 'York' 'East Toronto' 'West Toronto' 'Central Toronto' 'Mississauga']


### I choose downtown Toronto as my interest.

In [14]:
downtown_toronto = neighborhoods[neighborhoods['Borough'] == 'Downtown Toronto'].reset_index(drop=True)
downtown_toronto

Postal Code           Borough  \
0          M5A  Downtown Toronto   
1          M7A  Downtown Toronto   
2          M5B  Downtown Toronto   
3          M5C  Downtown Toronto   
4          M5E  Downtown Toronto   
5          M5G  Downtown Toronto   
6          M6G  Downtown Toronto   
7          M5H  Downtown Toronto   
8          M5J  Downtown Toronto   
9          M5K  Downtown Toronto   
10         M5L  Downtown Toronto   
11         M5S  Downtown Toronto   
12         M5T  Downtown Toronto   
13         M5V  Downtown Toronto   
14         M4W  Downtown Toronto   
15         M5W  Downtown Toronto   
16         M4X  Downtown Toronto   
17         M5X  Downtown Toronto   
18         M4Y  Downtown Toronto   

                                         Neighborhood   Latitude  Longitude  
0                           Regent Park, Harbourfront  43.654260 -79.360636  
1         Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
2                            Garden District, Ryerson  43.657162 -79.378937  
3                                      St. James Town  43.651494 -79.375418  
4                                         Berczy Park  43.644771 -79.373306  
5                                  Central Bay Street  43.657952 -79.387383  
6                                            Christie  43.669542 -79.422564  
7                            Richmond, Adelaide, King  43.650571 -79.384568  
8   Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752  
9            Toronto Dominion Centre, Design Exchange  43.647177 -79.381576  
10                     Commerce Court, Victoria Hotel  43.648198 -79.379817  
11                     University of Toronto, Harbord  43.662696 -79.400049  
12          Kensington Market, Chinatown, Grange Park  43.653206 -79.400049  
13  CN Tower, King and Spadina, Railway Lands, Har...  43.628947 -79.394420  
14                                           Rosedale  43.679563 -79.377529  
15                                     Stn A PO Boxes  43.646435 -79.374846  
16                        St. James Town, Cabbagetown  43.667967 -79.367675  
17             First Canadian Place, Underground city  43.648429 -79.382280  
18                               Church and Wellesley  43.665860 -79.383160

In [15]:
downtown_toronto.shape

(19, 5)

#### Create a map of all Toronto with neighborhoods superimposed on top.

In [16]:
# Location of University of Toronto
latitude = 43.662696
longitude = -79.400049

In [17]:
!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

Solving environment: - 
  - anaconda/osx-64::openssl-1.1.1d-h1de35cc_2
  - defaults/osx-64::openssl-1.1.1d-h1de35ccdone

# All requested packages already installed.



In [18]:
# create map of New York using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(neighborhoods['Latitude'], neighborhoods['Longitude'], neighborhoods['Borough'], neighborhoods['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

#### We are going to further create a map of Downtown Toronto with neighborhoods superimposed on top.

In [19]:
# create map of New York using latitude and longitude values
map_downtown_toronto = folium.Map(location=[latitude, longitude], zoom_start=12)

# add markers to map
for lat, lng, borough, neighborhood in zip(downtown_toronto['Latitude'], downtown_toronto['Longitude'], downtown_toronto['Borough'], downtown_toronto['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_downtown_toronto)  
    
map_downtown_toronto

Next, we are going to start utilizing the Foursquare API to explore the neighborhoods and segment them.

#### Define Foursquare Credentials and Version

In [20]:
CLIENT_ID = 'LSDMTKZ05ASUC4KBVXZWYLENV1TDP4DERWQPLP5AXB3W2YEH' # your Foursquare ID
CLIENT_SECRET = '5SLKOAU3MCFQVJTWYPMJGKVPCWC5VUZ1QOICSWIFMIDVYQQ2' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: LSDMTKZ05ASUC4KBVXZWYLENV1TDP4DERWQPLP5AXB3W2YEH
CLIENT_SECRET:5SLKOAU3MCFQVJTWYPMJGKVPCWC5VUZ1QOICSWIFMIDVYQQ2


#### Let's explore the first neighborhood in our dataframe.

In [21]:
# Get the neighborhood's name.
downtown_toronto.loc[11, 'Neighborhood']

'University of Toronto, Harbord'

In [22]:
# Get the neighborhood's latitude and longitude values.

neighborhood_latitude = downtown_toronto.loc[11, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = downtown_toronto.loc[11, 'Longitude'] # neighborhood longitude value

neighborhood_name = downtown_toronto.loc[11, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of University of Toronto, Harbord are 43.6626956, -79.4000493.


#### Now, let's get the top 100 venues that are in Marble Hill within a radius of 500 meters.
First, let's create the GET request URL. Name your URL url.

In [23]:
LIMIT = 100 # limit of number of venues returned by Foursquare API
radius = 500 # define radius
# create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url # display URL

'https://api.foursquare.com/v2/venues/explore?&client_id=LSDMTKZ05ASUC4KBVXZWYLENV1TDP4DERWQPLP5AXB3W2YEH&client_secret=5SLKOAU3MCFQVJTWYPMJGKVPCWC5VUZ1QOICSWIFMIDVYQQ2&v=20180605&ll=43.6626956,-79.4000493&radius=500&limit=100'

Send the GET request and examine the resutls

In [24]:
import requests # library to handle requests

results = requests.get(url).json()
#results

From the Foursquare lab in the previous module, we know that all the information is in the items key. Before we proceed, let's borrow the **get_category_type** function from the Foursquare lab.

In [25]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

Now we are ready to clean the json and structure it into a *pandas* dataframe.

In [26]:
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name           categories        lat        lng
0                              Yasu  Japanese Restaurant  43.662837 -79.403217
1                              Rasa           Restaurant  43.662757 -79.403988
2               The Dessert Kitchen         Dessert Shop  43.662823 -79.402746
3                       Piano Piano   Italian Restaurant  43.662949 -79.402898
4  Her Father's Cider Bar + Kitchen             Beer Bar  43.662448 -79.404703

In [27]:
nearby_venues.shape

(36, 4)

And how many venues were returned by Foursquare?

In [28]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

36 venues were returned by Foursquare.


## STEP 2. Explore Neighborhoods in Manhattan

#### Let's create a function to repeat the same process to all the neighborhoods in Manhattan

In [29]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

#### Now write the code to run the above function on each neighborhood and create a new dataframe called manhattan_venues.

In [30]:
downtown_toronto_venues = getNearbyVenues(names=downtown_toronto['Neighborhood'],
                                   latitudes=downtown_toronto['Latitude'],
                                   longitudes=downtown_toronto['Longitude']
                                  )

Regent Park, Harbourfront
Queen's Park, Ontario Provincial Government
Garden District, Ryerson
St. James Town
Berczy Park
Central Bay Street
Christie
Richmond, Adelaide, King
Harbourfront East, Union Station, Toronto Islands
Toronto Dominion Centre, Design Exchange
Commerce Court, Victoria Hotel
University of Toronto, Harbord
Kensington Market, Chinatown, Grange Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
Rosedale
Stn A PO Boxes
St. James Town, Cabbagetown
First Canadian Place, Underground city
Church and Wellesley


#### Let's check the size of the resulting dataframe

In [31]:
print(downtown_toronto_venues.shape)
downtown_toronto_venues.head()

(1213, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Regent Park, Harbourfront               43.65426              -79.360636   
1  Regent Park, Harbourfront               43.65426              -79.360636   
2  Regent Park, Harbourfront               43.65426              -79.360636   
3  Regent Park, Harbourfront               43.65426              -79.360636   
4  Regent Park, Harbourfront               43.65426              -79.360636   

                    Venue  Venue Latitude  Venue Longitude  \
0        Roselle Desserts       43.653447       -79.362017   
1           Tandem Coffee       43.653559       -79.361809   
2      Morning Glory Cafe       43.653947       -79.361149   
3  Cooper Koo Family YMCA       43.653249       -79.358008   
4     Body Blitz Spa East       43.654735       -79.359874   

        Venue Category  
0               Bakery  
1          Coffee Shop  
2       Breakfast Spot  
3  Distribution Center  
4                  Spa

Let's check how many venues were returned for each neighborhood

In [32]:
downtown_toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            54   
CN Tower, King and Spadina, Railway Lands, Harb...                     15   
Central Bay Street                                                     65   
Christie                                                               17   
Church and Wellesley                                                   77   
Commerce Court, Victoria Hotel                                        100   
First Canadian Place, Underground city                                100   
Garden District, Ryerson                                              100   
Harbourfront East, Union Station, Toronto Islands                     100   
Kensington Market, Chinatown, Grange Park                              56   
Queen's Park, Ontario Provincial Government                            36   
Regent Park, Harbourfront                                              46   
Richmond, Adelaide, King                                               92   
Rosedale                                                                4   
St. James Town                                                         77   
St. James Town, Cabbagetown                                            45   
Stn A PO Boxes                                                         93   
Toronto Dominion Centre, Design Exchange                              100   
University of Toronto, Harbord                                         36   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             54   
CN Tower, King and Spadina, Railway Lands, Harb...                      15   
Central Bay Street                                                      65   
Christie                                                                17   
Church and Wellesley                                                    77   
Commerce Court, Victoria Hotel                                         100   
First Canadian Place, Underground city                                 100   
Garden District, Ryerson                                               100   
Harbourfront East, Union Station, Toronto Islands                      100   
Kensington Market, Chinatown, Grange Park                               56   
Queen's Park, Ontario Provincial Government                             36   
Regent Park, Harbourfront                                               46   
Richmond, Adelaide, King                                                92   
Rosedale                                                                 4   
St. James Town                                                          77   
St. James Town, Cabbagetown                                             45   
Stn A PO Boxes                                                          93   
Toronto Dominion Centre, Design Exchange                               100   
University of Toronto, Harbord                                          36   

                                                    Venue  Venue Latitude  \
Neighborhood                                                                
Berczy Park                                            54              54   
CN Tower, King and Spadina, Railway Lands, Harb...     15              15   
Central Bay Street                                     65              65   
Christie                                               17              17   
Church and Wellesley                                   77              77   
Commerce Court, Victoria Hotel                        100             100   
First Canadian Place, Underground city                100             100   
Garden District, Ryerson                              100             100   
Harbourfront East, Union 

#### Let's find out how many unique categories can be curated from all the returned venues

In [33]:
print('There are {} uniques categories.'.format(len(downtown_toronto_venues['Venue Category'].unique())))

There are 210 uniques categories.


## STEP 3. Analyze Each Neighborhood

In [34]:
downtown_toronto_venues.head()

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Regent Park, Harbourfront               43.65426              -79.360636   
1  Regent Park, Harbourfront               43.65426              -79.360636   
2  Regent Park, Harbourfront               43.65426              -79.360636   
3  Regent Park, Harbourfront               43.65426              -79.360636   
4  Regent Park, Harbourfront               43.65426              -79.360636   

                    Venue  Venue Latitude  Venue Longitude  \
0        Roselle Desserts       43.653447       -79.362017   
1           Tandem Coffee       43.653559       -79.361809   
2      Morning Glory Cafe       43.653947       -79.361149   
3  Cooper Koo Family YMCA       43.653249       -79.358008   
4     Body Blitz Spa East       43.654735       -79.359874   

        Venue Category  
0               Bakery  
1          Coffee Shop  
2       Breakfast Spot  
3  Distribution Center  
4                  Spa

In [35]:
downtown_toronto_venues.shape

(1213, 7)

In [36]:
# one hot encoding
downtown_toronto_onehot = pd.get_dummies(downtown_toronto_venues[['Venue Category']], prefix="", prefix_sep="")
downtown_toronto_onehot.shape

(1213, 210)

In [37]:
downtown_toronto_onehot.head()

Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0                  0        0                   0             0   
1                  0        0                   0             0   
2                  0        0                   0             0   
3                  0        0                   0             0   
4                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  ...  Theme Restaurant  Trail  Train Station  \
0             0         0  ...                 0      0              0   
1             0         0  ...                 0      0              0   
2             0         0  ...                 0      0              0   
3             0         0  ...                 0      0              0   
4             0         0  ...                 0      0              0   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                              0                 0                      0   
1                              0                 0                      0   
2                              0                 0                      0   
3                              0                 0                      0   
4                              0                 0                      0   

   Wine Bar  Wings Joint  Women's Store  Yoga Studio  
0         0            0              0            0  
1         0            0              0            0  
2         0            0              0            0  
3         0            0              0            0  
4         0            0              0            0  

[5 rows x 210 columns]

In [38]:
# add neighborhood column back to dataframe
downtown_toronto_onehot['Neighborhood'] = downtown_toronto_venues['Neighborhood']
downtown_toronto_onehot.shape

(1213, 210)

In [39]:
downtown_toronto_onehot.head()

Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0                  0        0                   0             0   
1                  0        0                   0             0   
2                  0        0                   0             0   
3                  0        0                   0             0   
4                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  ...  Theme Restaurant  Trail  Train Station  \
0             0         0  ...                 0      0              0   
1             0         0  ...                 0      0              0   
2             0         0  ...                 0      0              0   
3             0         0  ...                 0      0              0   
4             0         0  ...                 0      0              0   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                              0                 0                      0   
1                              0                 0                      0   
2                              0                 0                      0   
3                              0                 0                      0   
4                              0                 0                      0   

   Wine Bar  Wings Joint  Women's Store  Yoga Studio  
0         0            0              0            0  
1         0            0              0            0  
2         0            0              0            0  
3         0            0              0            0  
4         0            0              0            0  

[5 rows x 210 columns]

In [40]:
#######
# define a list of column names
cols = downtown_toronto_onehot.columns.tolist()
cols

# move the column name to the beggining
cols.insert(0, cols.pop(cols.index('Neighborhood')))
cols

#then use .reindex() function to reorder
downtown_toronto_onehot = downtown_toronto_onehot.reindex(columns= cols)

#check result
downtown_toronto_onehot.head()

Neighborhood  Afghan Restaurant  Airport  Airport Food Court  \
0  Regent Park, Harbourfront                  0        0                   0   
1  Regent Park, Harbourfront                  0        0                   0   
2  Regent Park, Harbourfront                  0        0                   0   
3  Regent Park, Harbourfront                  0        0                   0   
4  Regent Park, Harbourfront                  0        0                   0   

   Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0             0               0                0                 0   
1             0               0                0                 0   
2             0               0                0                 0   
3             0               0                0                 0   
4             0               0                0                 0   

   American Restaurant  Antique Shop  ...  Theme Restaurant  Trail  \
0                    0             0  ...                 0      0   
1                    0             0  ...                 0      0   
2                    0             0  ...                 0      0   
3                    0             0  ...                 0      0   
4                    0             0  ...                 0      0   

   Train Station  Vegetarian / Vegan Restaurant  Video Game Store  \
0              0                              0                 0   
1              0                              0                 0   
2              0                              0                 0   
3              0                              0                 0   
4              0                              0                 0   

   Vietnamese Restaurant  Wine Bar  Wings Joint  Women's Store  Yoga Studio  
0                      0         0            0              0            0  
1                      0         0            0              0            0  
2                      0         0            0              0            0  
3                      0         0            0              0            0  
4                      0         0            0              0            0  

[5 rows x 210 columns]

In [41]:
downtown_toronto_onehot.shape

(1213, 210)

#### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [42]:
downtown_toronto_grouped = downtown_toronto_onehot.groupby('Neighborhood').mean().reset_index()
downtown_toronto_grouped

Neighborhood  Afghan Restaurant  \
0                                         Berczy Park           0.000000   
1   CN Tower, King and Spadina, Railway Lands, Har...           0.000000   
2                                  Central Bay Street           0.000000   
3                                            Christie           0.000000   
4                                Church and Wellesley           0.012987   
5                      Commerce Court, Victoria Hotel           0.000000   
6              First Canadian Place, Underground city           0.000000   
7                            Garden District, Ryerson           0.000000   
8   Harbourfront East, Union Station, Toronto Islands           0.000000   
9           Kensington Market, Chinatown, Grange Park           0.000000   
10        Queen's Park, Ontario Provincial Government           0.000000   
11                          Regent Park, Harbourfront           0.000000   
12                           Richmond, Adelaide, King           0.000000   
13                                           Rosedale           0.000000   
14                                     St. James Town           0.000000   
15                        St. James Town, Cabbagetown           0.000000   
16                                     Stn A PO Boxes           0.000000   
17           Toronto Dominion Centre, Design Exchange           0.000000   
18                     University of Toronto, Harbord           0.000000   

     Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0   0.000000            0.000000      0.000000        0.000000   
1   0.066667            0.066667      0.066667        0.066667   
2   0.000000            0.000000      0.000000        0.000000   
3   0.000000            0.000000      0.000000        0.000000   
4   0.000000            0.000000      0.000000        0.000000   
5   0.000000            0.000000      0.000000        0.000000   
6   0.000000            0.000000      0.000000        0.000000   
7   0.000000            0.000000      0.000000        0.000000   
8   0.000000            0.000000      0.000000        0.000000   
9   0.000000            0.000000      0.000000        0.000000   
10  0.000000            0.000000      0.000000        0.000000   
11  0.000000            0.000000      0.000000        0.000000   
12  0.000000            0.000000      0.000000        0.000000   
13  0.000000            0.000000      0.000000        0.000000   
14  0.000000            0.000000      0.000000        0.000000   
15  0.000000            0.000000      0.000000        0.000000   
16  0.000000            0.000000      0.000000        0.000000   
17  0.000000            0.000000      0.000000        0.000000   
18  0.000000            0.000000      0.000000        0.000000   

    Airport Service  Airport Terminal  American Restaurant  Antique Shop  ...  \
0               0.0          0.000000             0.000000      0.000000  ...   
1               0.2          0.133333             0.000000      0.000000  ...   
2               0.0          0.000000             0.000000      0.000000  ...   
3               0.0          0.000000             0.000000      0.000000  ...   
4               0.0          0.000000             0.012987      0.000000  ...   
5               0.0          0.000000             0.040000      0.000000  ...   
6               0.0          0.000000             0.030000      0.000000  ...   
7               0.0          0.000000             0.000000      0.000000  ...   
8               0.0          0.000000             0.000000      0.000000  ...   
9               0.0          0.000000             0.000000      0.000000  ...   
10              0.0          0.000000             0.000000      0.000000  ...   
11              0.0          0.000000             0.000000      0.021739  ...   
12              0.0          0.000000             0.021739      0.000000  ...   
13              0.0          0.000000             0.000000      0

#### Let's confirm the new size

In [43]:
downtown_toronto_grouped.shape

(19, 210)

#### Let's print each neighborhood along with the top 5 most common venues

In [44]:
num_top_venues = 5

for hood in downtown_toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = downtown_toronto_grouped[downtown_toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
                venue  freq
0         Coffee Shop  0.07
1        Cocktail Bar  0.06
2            Beer Bar  0.04
3                Café  0.04
4  Seafood Restaurant  0.04


----CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport----
                 venue  freq
0      Airport Service  0.20
1     Airport Terminal  0.13
2        Boat or Ferry  0.07
3      Harbor / Marina  0.07
4  Rental Car Location  0.07


----Central Bay Street----
                 venue  freq
0          Coffee Shop  0.17
1       Sandwich Place  0.06
2   Italian Restaurant  0.06
3                 Café  0.05
4  Japanese Restaurant  0.03


----Christie----
           venue  freq
0  Grocery Store  0.24
1           Café  0.18
2           Park  0.12
3          Diner  0.06
4     Restaurant  0.06


----Church and Wellesley----
                 venue  freq
0          Coffee Shop  0.08
1  Japanese Restaurant  0.06
2     Sushi Restaurant  0.06
3           Re

#### Let's put that into a *pandas* dataframe
First, let's write a function to sort the venues in descending order.

In [45]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [46]:
import numpy as np # library to handle data in a vectorized manner

num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = downtown_toronto_grouped['Neighborhood']

for ind in np.arange(downtown_toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(downtown_toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1  CN Tower, King and Spadina, Railway Lands, Har...       Airport Service   
2                                 Central Bay Street           Coffee Shop   
3                                           Christie         Grocery Store   
4                               Church and Wellesley           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0          Cocktail Bar                  Café              Beer Bar   
1      Airport Terminal              Boutique   Rental Car Location   
2    Italian Restaurant        Sandwich Place                  Café   
3                  Café                  Park    Athletics & Sports   
4      Sushi Restaurant   Japanese Restaurant               Gay Bar   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0            Restaurant    Seafood Restaurant           Cheese Shop   
1      Sculpture Garden                 Plane       Harbor / Marina   
2           Salad Place                   Bar       Bubble Tea Shop   
3    Italian Restaurant                 Diner             Nightclub   
4            Restaurant                  Café                 Hotel   

  8th Most Common Venue    9th Most Common Venue    10th Most Common Venue  
0                Bakery  Comfort Food Restaurant        Basketball Stadium  
1         Boat or Ferry           Airport Lounge              Airport Gate  
2          Burger Joint      Japanese Restaurant            Ice Cream Shop  
3           Candy Store               Restaurant                Baby Store  
4             Gastropub              Yoga Studio  Mediterranean Restaurant

## STEP 4. Cluster Neighborhoods

Run k-means to cluster the neighborhood into 3 clusters.

In [47]:
# import k-means from clustering stage
from sklearn.cluster import KMeans

In [48]:
# set number of clusters
kclusters = 5
downtown_toronto_grouped_clustering = downtown_toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(downtown_toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_

array([0, 1, 0, 4, 0, 0, 0, 0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 0, 3],
      dtype=int32)

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [49]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

downtown_toronto_merged = downtown_toronto

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
downtown_toronto_merged = downtown_toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

downtown_toronto_merged.head() # check the last columns!

Postal Code           Borough                                 Neighborhood  \
0         M5A  Downtown Toronto                    Regent Park, Harbourfront   
1         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2         M5B  Downtown Toronto                     Garden District, Ryerson   
3         M5C  Downtown Toronto                               St. James Town   
4         M5E  Downtown Toronto                                  Berczy Park   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.654260 -79.360636               0           Coffee Shop   
1  43.662301 -79.389494               0           Coffee Shop   
2  43.657162 -79.378937               0        Clothing Store   
3  43.651494 -79.375418               0                  Café   
4  43.644771 -79.373306               0           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                   Pub                  Park                Bakery   
1      Sushi Restaurant     College Cafeteria           Yoga Studio   
2           Coffee Shop       Bubble Tea Shop        Cosmetics Shop   
3           Coffee Shop          Cocktail Bar             Gastropub   
4          Cocktail Bar                  Café              Beer Bar   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0        Breakfast Spot                  Café               Theater   
1           Music Venue              Beer Bar        Sandwich Place   
2                  Café   Japanese Restaurant    Italian Restaurant   
3   American Restaurant               Theater            Restaurant   
4            Restaurant    Seafood Restaurant           Cheese Shop   

       8th Most Common Venue    9th Most Common Venue 10th Most Common Venue  
0                      Hotel        French Restaurant      Health Food Store  
1               Burger Joint            Burrito Place                   Café  
2  Middle Eastern Restaurant               Restaurant              Bookstore  
3                        Gym                 Beer Bar       Department Store  
4                     Bakery  Comfort Food Restaurant     Basketball Stadium

Finally, let's visualize the resulting clusters

In [50]:
# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

In [51]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(downtown_toronto_merged['Latitude'], downtown_toronto_merged['Longitude'], downtown_toronto_merged['Neighborhood'], downtown_toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## STEP 5. Examine Clusters

Now, you can examine each cluster and determine the discriminating venue categories that distinguish each cluster. 

In [52]:
# Cluster 1
downtown_toronto_merged.loc[downtown_toronto_merged['Cluster Labels'] == 0, downtown_toronto_merged.columns[[2] + list(range(6, downtown_toronto_merged.shape[1]))]]

Neighborhood 1st Most Common Venue  \
0                           Regent Park, Harbourfront           Coffee Shop   
1         Queen's Park, Ontario Provincial Government           Coffee Shop   
2                            Garden District, Ryerson        Clothing Store   
3                                      St. James Town                  Café   
4                                         Berczy Park           Coffee Shop   
5                                  Central Bay Street           Coffee Shop   
7                            Richmond, Adelaide, King           Coffee Shop   
8   Harbourfront East, Union Station, Toronto Islands           Coffee Shop   
9            Toronto Dominion Centre, Design Exchange           Coffee Shop   
10                     Commerce Court, Victoria Hotel           Coffee Shop   
12          Kensington Market, Chinatown, Grange Park                  Café   
15                                     Stn A PO Boxes           Coffee Shop   
16                        St. James Town, Cabbagetown            Restaurant   
17             First Canadian Place, Underground city           Coffee Shop   
18                               Church and Wellesley           Coffee Shop   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                     Pub                  Park                Bakery   
1        Sushi Restaurant     College Cafeteria           Yoga Studio   
2             Coffee Shop       Bubble Tea Shop        Cosmetics Shop   
3             Coffee Shop          Cocktail Bar             Gastropub   
4            Cocktail Bar                  Café              Beer Bar   
5      Italian Restaurant        Sandwich Place                  Café   
7                    Café            Restaurant        Clothing Store   
8                Aquarium                  Café                 Hotel   
9                   Hotel                  Café            Restaurant   
10                   Café            Restaurant                 Hotel   
12  Vietnamese Restaurant                Bakery    Mexican Restaurant   
15                   Café    Seafood Restaurant            Restaurant   
16            Coffee Shop    Chinese Restaurant                  Park   
17                   Café                   Gym   Japanese Restaurant   
18       Sushi Restaurant   Japanese Restaurant               Gay Bar   

   5th Most Common Venue 6th Most Common Venue          7th Most Common Venue  \
0         Breakfast Spot                  Café                        Theater   
1            Music Venue              Beer Bar                 Sandwich Place   
2                   Café   Japanese Restaurant             Italian Restaurant   
3    American Restaurant               Theater                     Restaurant   
4             Restaurant    Seafood Restaurant                    Cheese Shop   
5            Salad Place                   Bar                Bubble Tea Shop   
7                  Hotel       Thai Restaurant                            Gym   
8                Brewery            Restaurant            Sporting Goods Shop   
9    American Restaurant           Salad Place            Japanese Restaurant   
10   American Restaurant                   Gym                  Deli / Bodega   
12           Coffee Shop           Gaming Cafe  Vegetarian / Vegan Restaurant   
15          Cocktail Bar   Japanese Restaurant             Italian Restaurant   
16                   Pub    Italian Restaurant                         Bakery   
17                 Hotel            Restaurant             Seafood Restaurant   
18            Restaurant                  Café                          Hotel   

        8th Most Common Venue    9th Most Common Venue  \
0                       Hotel        French Restaurant   
1                Burger Joint            Burrito Place   
2   Middle Eastern Restaurant               Restaurant   
3                         Gym                 Beer Bar   
4                   

In [53]:
# Cluster 2
downtown_toronto_merged.loc[downtown_toronto_merged['Cluster Labels'] == 1, downtown_toronto_merged.columns[[2] + list(range(6, downtown_toronto_merged.shape[1]))]]

Neighborhood 1st Most Common Venue  \
13  CN Tower, King and Spadina, Railway Lands, Har...       Airport Service   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
13      Airport Terminal              Boutique   Rental Car Location   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
13      Sculpture Garden                 Plane       Harbor / Marina   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
13         Boat or Ferry        Airport Lounge           Airport Gate

In [54]:
# Cluster 3
downtown_toronto_merged.loc[downtown_toronto_merged['Cluster Labels'] == 2, downtown_toronto_merged.columns[[2] + list(range(6, downtown_toronto_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
14     Rosedale                  Park            Playground   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
14                 Trail              Creperie      Doner Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
14               Dog Run   Distribution Center        Discount Store   

   9th Most Common Venue 10th Most Common Venue  
14                 Diner           Dessert Shop

In [55]:
# Cluster 4
downtown_toronto_merged.loc[downtown_toronto_merged['Cluster Labels'] == 3, downtown_toronto_merged.columns[[2] + list(range(6, downtown_toronto_merged.shape[1]))]]

Neighborhood 1st Most Common Venue  \
11  University of Toronto, Harbord                  Café   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
11            Restaurant                   Bar    Italian Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
11   Japanese Restaurant             Bookstore                Bakery   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
11        Sandwich Place              Beer Bar             Beer Store

In [56]:
# Cluster 5
downtown_toronto_merged.loc[downtown_toronto_merged['Cluster Labels'] == 4, downtown_toronto_merged.columns[[2] + list(range(6, downtown_toronto_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
6     Christie         Grocery Store                  Café   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
6                  Park    Athletics & Sports    Italian Restaurant   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
6                 Diner             Nightclub           Candy Store   

  9th Most Common Venue 10th Most Common Venue  
6            Restaurant             Baby Store

## I clustered those neighborhoods into 5 groups. 